CitadelExtras/HeadBob.cs

```diff
@@ -0,0 +1,92 @@
+﻿/*
+WIP Headbobbing, weapon bobbing, and footsteps NOV 22.
+
+IMPORTANT: In order for headbobbing to work, the Y value "0.84f" for the player MUST be removed in the appropriate script.
+*/
+
+using UnityEngine;
+
+[System.Serializable]
+public class HeadBob : MonoBehaviour
+{
+    
+    public GameObject capsule = null;
+    public Rigidbody rb = null;
+
+    public GameObject Pipe;
+    public float bobRate = 10.0f;
+    public float bob = 0.02f;
+
+    public float dHeight = 0.84f;
+    public float stepRate = 0.5f;
+    public float stepCoolDown;
+
+    AudioSource audio;
+    public AudioClip[] steps;
+    
+    float timer = 0;
+
+    
+    void Start()
+    {
+        steps = new AudioClip[4];
+        capsule = GameObject.Find("PlayerCapsule");
+        rb = capsule.GetComponent<Rigidbody>();
+        audio = capsule.GetComponent<AudioSource>();
+        Pipe = GameObject.Find("PipeTransformOffset");
+    }
+
+    
+    void Update()
+    {
+        if(Mathf.Abs(rb.velocity.x) > 0.1f || Mathf.Abs(rb.velocity.z) > 0.1f )
+        {
+            //Player is moving
+            timer += Time.deltaTime * bobRate;
+            transform.localPosition = new Vector3(
+
+                Mathf.Sin(timer) * bob, 
+                dHeight + -(Mathf.Sin(timer)) * bob, 
+                transform.localPosition.z
+
+                );

```

Comment:
The result of this is such that the view bobs up and to the left then down and to the right in a diagonal pattern.  Should probably have the y flip sign when x is near zero when moving toward zero (like within 0.01 or so, so it doesn't miss detecting the zero crossing) so that it does more of a smile shape by going up and to the left then back to center then up and to the right.  At the moment it feels like the player is limping (might be kind of cool if health is really low though haha).

Should take the bodystate into account as well to modulate the strength depending on if crouched vs prone vs standing so that it barely bobs at all while crawling.  Actually, would make more sense to be more side to side when crawling with almost no up/down.

Reply:
Side note, the original game applies some slight lean in the direction of travel.  Would be interesting to investigate adding that to PlayerMovement in the Lean function.  Not sure how I overlooked adding that.

Reply:
Also, in my testing I moved this into MouseLookScript.cs inside Recoil to avoid potential for race conditions and have recoil override the bob.  Would be better to do that rather than have this as a separate script; I strive to make sure that all script actions that affect a particular transform only happen in one place as I've ran into race conditions causing jitter and other errors.

Reply:
Mostly just noting things for myself at this point, sorry for Pacmikey or whoever may be following along.

I just fixed a bug with the camera height such that prone and crouch were broken as a result of not applying the PlayerMovement.a.currentCrouchRatio scale to the 0.84f standing camera height.  Without this applied, the camera was floating just above the capsule collider allowing the player to jump and ghost the camera through the ceiling.  Worse yet, the CantStand/CantCrouch functions were relying on camera position and giving false negatives as a result.  This is now fixed (well will be once I push the commit here in a few minutes) and will change how view bobbing will ultimately be implemented.

As an aside, I think it would behoove us/me to set a focal point 20.0f or so in front of and attached to the LeanCapsule that the player camera should be pointed to via the lookto function Unity's API provides.  This will help to prevent any camera bobbing from being too jarring as what the player may reasonably be assumed to be looking at will be kept more or less in the center of the screen.  This seems to be a common practice in other FPS games, such as Mirror's Edge.  Also the strength here should be reduced by 25 to 33% or so.  It's a tad much.

I'll need to add a toggle in Config options to allow this to be disabled but doesn't have to be done here in this PR.  And may end up implementing this in the way that will work best with prone crouching myself and close this PR thereafter without merging it.

Reply:
nah I get it, you and I are the only people looking at this Github so it would get kinda lonely with no one to update to.

Reply:
I got so depressed seeing this awesome project have like, no one contribute to it other than like you and 1 or 2 other people